src/common_modules/vsts-cli-package-common/vsts/cli/package/common/artifacttool.py

```diff
@@ -0,0 +1,97 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import json
+import logging
+import os
+
+from knack.util import CLIError
+
+from vsts.cli.common.services import _get_credentials
+
+logger = logging.getLogger('vsts.packaging')
+
+class ArtifactToolInvoker:
+    def __init__(self, tool_invoker, artifacttool_updater):
+        self._tool_invoker = tool_invoker
+        self._artifacttool_updater = artifacttool_updater
+
+    PATVAR = "VSTS_ARTIFACTTOOL_PATVAR"
+
+    def download_upack(self, team_instance, feed, package_name, package_version, path):
+        self.run_artifacttool(team_instance, ["upack", "download", "--service", team_instance, "--patvar", self.PATVAR, "--feed", feed, "--package-name", package_name, "--package-version", package_version, "--path", path], "Downloading")
+
+    def publish_upack(self, team_instance, feed, package_name, package_version, description, path):
+        args = ["upack", "publish", "--service", team_instance, "--patvar", self.PATVAR, "--feed", feed, "--package-name", package_name, "--package-version", package_version, "--path", path]
+        if description:
+            args.extend(["--description", description])
+        self.run_artifacttool(team_instance, args, "Publishing")
+
+    def run_artifacttool(self, team_instance, args, initial_progress_message):
+        # Download ArtifactTool if necessary, and return the path
+        artifacttool_dir = self._artifacttool_updater.get_latest_artifacttool(team_instance)
+        artifacttool_binary_path = os.path.join(artifacttool_dir, "ArtifactTool")
+
+        # Populate the environment for the process with the PAT
+        creds = _get_credentials(team_instance)
+        new_env = os.environ.copy()
+        new_env[self.PATVAR] = str(creds.password)
+
+        # Run ArtifactTool
+        command_args = [artifacttool_binary_path] + args
+        proc = self._tool_invoker.run(command_args, new_env, initial_progress_message, self._process_stderr)
+        print(proc.stdout.read().decode('utf-8'))
+
+    def _process_stderr(self, line, update_progress_callback):
+        try:
+            json_line = json.loads(line)
+        except Exception as ex:
+            json_line = None
+            logger.warning("Failed to parse structured output from UPack tooling (ArtifactTool)")
+            logger.warning("Exception: %s" % ex)
```

Comment:
The Internet seems to favor .format() rather than the modulo operator for string formatting. Is this style used in other places of the CLI; if it's not, consider changing to .format

Reply:
Will change.